WcaOnRails/lib/competition_results_validator.rb

```diff
@@ -0,0 +1,606 @@
+      "results" => {
+        "type" => "array",
+        "items" => INDIVIDUAL_RESULT_JSON_SCHEMA,
+      },
+      "groups" => {
+        "type" => "array",
+        "items" => GROUP_JSON_SCHEMA,
+      },
+    },
+    "required" => ["roundId", "formatId", "results", "groups"],
+  }.freeze
+
+  PERSON_JSON_SCHEMA = {
+    "type" => "object",
+    "properties" => {
+      "id" => { "type" => "number" },
+      "name" => { "type" => "string" },
+      # May be empty
+      "wcaId" => { "type" => "string" },
+      "countryId" => { "type" => "string" },
+      # May be empty
+      "gender" => { "type" => "string" },
+      "dob" => { "type" => "string" },
+    },
+    "required" => ["id", "name", "wcaId", "countryId", "gender", "dob"],
+  }.freeze
+
+  EVENT_JSON_SCHEMA = {
+    "type" => "object",
+    "properties" => {
+      "eventId" => { "type" => "string" },
+      "rounds" => {
+        "type" => "array",
+        "items" => ROUND_JSON_SCHEMA,
+      },
+    },
+    "required" => ["eventId", "rounds"],
+  }.freeze
+
+  RESULT_JSON_SCHEMA = {
+    "type" => "object",
+    "properties" => {
+      "formatVersion" => { "type" => "string" },
+      "competitionId" => { "type" => "string" },
+      "persons" => {
+        "type" => "array",
+        "items" => PERSON_JSON_SCHEMA,
+      },
+      "events" => {
+        "type" => "array",
+        "items" => EVENT_JSON_SCHEMA,
+      },
+    },
+    "required" => ["formatVersion", "competitionId", "persons", "events"],
+  }.freeze
+
+  def initialize(competition_id, check_real_results = false)
+    @errors = {
+      persons: [],
+      events: [],
+      rounds: [],
+      results: [],
+      scrambles: [],
+    }
+    @warnings = {
+      persons: [],
+      results: [],
+      rounds: [],
+    }
+    @total_errors = 0
+    @total_warnings = 0
+    @number_of_non_matching_rounds = 0
+
+    associations = {
+      events: [],
+      competition_events: {
+        rounds: [:competition_event, :format],
+      },
+    }
+
+    @competition = Competition.includes(associations).find(competition_id)
+
+    @check_real_results = check_real_results
+
+    result_model = @check_real_results ? Result : InboxResult
+    @results = result_model.sorted_for_competition(competition_id)
+    @has_results = @results.any?
+    unless @has_results
+      @total_errors = 1
+      @errors[:results] << "The competition has no result."
+      return
+    end
+
+    @persons = if @check_real_results
+                 @competition.competitors
+               else
+                 InboxPerson.where(competitionId: competition_id)
+               end
+
+    @scrambles = Scramble.where(competitionId: competition_id)
+
+    # check persons
+    # basic checks on persons are done in the model, uniqueness for a given competition
+    # is done in the SQL schema.
+
+    # Map a personId to its corresponding object
+    @persons_by_id = Hash[@persons.map { |person| [@check_real_results ? person.wca_id : person.id, person] }]
```

Comment:
Because we may deal with `Person`s from either the `InboxPerson` table (in which not everyone has a WCA ID, but everyone is indexed by their `personId`), or from the `Person` table (in which everyone has a WCA ID, but no `personId` for the competition).
(See line 186 above for the two possible cases)